File: IO.Astrodynamics.Net/IO.Astrodynamics/Body/Spacecraft/Payload.cs
using System;

namespace IO.Astrodynamics.Body.Spacecraft
{
    public class Payload : IEquatable<Payload>
    {
        
        public string Name { get; }
        public string SerialNumber { get; }
        public double Mass { get; }

        public Payload(string name, double mass, string serialNumber)
        {
            if (mass <= 0) throw new ArgumentException("Payload must have a mass");


            if (string.IsNullOrEmpty(name)) throw new ArgumentException("Value cannot be null or empty.", nameof(name));
            if (string.IsNullOrEmpty(serialNumber)) throw new ArgumentException("Value cannot be null or empty.", nameof(serialNumber));

            Name = name;
            Mass = mass;
            SerialNumber = serialNumber;
        }
        
        public bool Equals(Payload other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return Name == other.Name && SerialNumber == other.SerialNumber && Mass.Equals(other.Mass);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Payload)obj);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Name, SerialNumber, Mass);
        }

        public static bool operator ==(Payload left, Payload right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(Payload left, Payload right)
        {
            return !Equals(left, right);
        }

    }
}

File: IO.Astrodynamics/Body/Spacecraft/Payload.cpp
/*
 Copyright (c) 2021-2023. Sylvain Guillet (sylvain.guillet@tutamail.com)
 */
#include <Payload.h>
#include <InvalidArgumentException.h>

IO::Astrodynamics::Body::Spacecraft::Payload::Payload(const std::string &serialNumber, const std::string &name, const double mass)
{
    if (name.empty())
    {
        throw IO::Astrodynamics::Exception::InvalidArgumentException("Payload must have a name");
    }

    if (serialNumber.empty())
    {
        throw IO::Astrodynamics::Exception::InvalidArgumentException("Payload must have a serial number");
    }

    if (mass <= 0)
    {
        throw IO::Astrodynamics::Exception::InvalidArgumentException("Payload must have a positive mass");
    }
    
    const_cast<std::string &>(m_name) = name;
    const_cast<std::string &>(m_serialNumber) = serialNumber;
    const_cast<double &>(m_mass) = mass;
}

double IO::Astrodynamics::Body::Spacecraft::Payload::GetMass() const
{
    return m_mass;
}

std::string IO::Astrodynamics::Body::Spacecraft::Payload::GetName() const
{
    return m_name;
}

std::string IO::Astrodynamics::Body::Spacecraft::Payload::GetSerialNumber() const
{
    return m_serialNumber;
}
